source/opt/constants.cpp

```diff
@@ -0,0 +1,215 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "constants.h"
+#include "ir_context.h"
+
+#include <unordered_map>
+#include <vector>
+
+namespace spvtools {
+namespace opt {
+namespace analysis {
+
+analysis::Type* ConstantManager::GetType(const ir::Instruction* inst) const {
+  return context()->get_type_mgr()->GetType(inst->type_id());
+}
+
+uint32_t ConstantManager::FindRecordedConstant(
+    const analysis::Constant* c) const {
+  auto iter = const_val_to_id_.find(c);
+  if (iter == const_val_to_id_.end()) {
+    return 0;
+  } else {
+    return iter->second;
+  }
+}
+
+std::vector<const analysis::Constant*> ConstantManager::GetConstantsFromIds(
+    const std::vector<uint32_t>& ids) const {
+  std::vector<const analysis::Constant*> constants;
+  for (uint32_t id : ids) {
+    if (analysis::Constant* c = FindRecordedConstant(id)) {
+      constants.push_back(c);
+    } else {
+      return {};
+    }
+  }
+  return constants;
+}
+
+ir::Instruction* ConstantManager::BuildInstructionAndAddToModule(
+    std::unique_ptr<analysis::Constant> c, ir::Module::inst_iterator* pos) {
+  analysis::Constant* new_const = c.get();
+  uint32_t new_id = context()->TakeNextId();
+  const_val_to_id_[new_const] = new_id;
+  id_to_const_val_[new_id] = std::move(c);
+  auto new_inst = CreateInstruction(new_id, new_const);
+  if (!new_inst) return nullptr;
+  auto* new_inst_ptr = new_inst.get();
+  *pos = pos->InsertBefore(std::move(new_inst));
+  ++(*pos);
+  context()->get_def_use_mgr()->AnalyzeInstDefUse(new_inst_ptr);
+  return new_inst_ptr;
+}
+
+analysis::Constant* ConstantManager::FindRecordedConstant(uint32_t id) const {
+  auto iter = id_to_const_val_.find(id);
+  if (iter == id_to_const_val_.end()) {
+    return nullptr;
+  } else {
+    return iter->second.get();
+  }
+}
+
+std::unique_ptr<analysis::Constant> ConstantManager::CreateConstant(
+    const analysis::Type* type,
+    const std::vector<uint32_t>& literal_words_or_ids) const {
+  std::unique_ptr<analysis::Constant> new_const;
+  if (literal_words_or_ids.size() == 0) {
+    // Constant declared with OpConstantNull
+    return MakeUnique<analysis::NullConstant>(type);
+  } else if (auto* bt = type->AsBool()) {
+    assert(literal_words_or_ids.size() == 1 &&
+           "Bool constant should be declared with one operand");
+    return MakeUnique<analysis::BoolConstant>(bt, literal_words_or_ids.front());
+  } else if (auto* it = type->AsInteger()) {
+    return MakeUnique<analysis::IntConstant>(it, literal_words_or_ids);
+  } else if (auto* ft = type->AsFloat()) {
+    return MakeUnique<analysis::FloatConstant>(ft, literal_words_or_ids);
+  } else if (auto* vt = type->AsVector()) {
+    auto components = GetConstantsFromIds(literal_words_or_ids);
+    if (components.empty()) return nullptr;
```

Comment:
I'm trying to understand the uses cases.  Why return nullptr instead of asserting?  Do we envision a situation where the caller wants a constant vector, and can do something meaningful with the nullptr?  The only things I can think of is that the caller made a mistake.  They have an id for a non-constant, or they did not register the constant.

Reply:
This is code from the spec constant folder.  Any functionality changes I need in this code will come from CCP.